oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/OAuth2ClientCredentialsGrantBuilder.java

```diff
@@ -0,0 +1,156 @@
+     * authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     * @param authorizationType One of the registered HTTP authentication schemes as per
+     *                          <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">
+     *                          HTTP Authentication Scheme Registry</a>.
+     */
+    @Override
+    protected OAuth2ClientCredentialsGrantBuilder clientCredentials(
+            Supplier<Map.Entry<String, String>> credentialsSupplier, String authorizationType) {
+        super.clientCredentials(credentialsSupplier, authorizationType);
+        return this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on client credentials and
+     * {@code Basic} authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     */
+    @Override
+    protected OAuth2ClientCredentialsGrantBuilder clientCredentials(
+            Supplier<Map.Entry<String, String>> credentialsSupplier) {
+        super.clientCredentials(credentialsSupplier);
+        return this;
+    }
+
+    /**
+     * A period when the token should be refreshed proactively prior to its expiry.
+     */
+    @Override
+    protected OAuth2ClientCredentialsGrantBuilder refreshBefore(Duration refreshBefore) {
+        super.refreshBefore(refreshBefore);
+        return this;
+    }
+
+    /**
+     * A {@link Supplier} to load Access Token from, to be able to restore the previous session. OPTIONAL.
+     */
+    @Override
+    protected OAuth2ClientCredentialsGrantBuilder tokenSupplier(Supplier<AccessTokenCapsule> tokenSupplier) {
+        super.tokenSupplier(tokenSupplier);
+        return this;
+    }
+
+    /**
+     * A {@link Consumer} to store Access Token to, to be able restore the previous session. OPTIONAL.
+     */
+    @Override
+    protected OAuth2ClientCredentialsGrantBuilder tokenConsumer(Consumer<AccessTokenCapsule> tokenConsumer) {
+        super.tokenConsumer(tokenConsumer);
+        return this;
+    }
+
+    @Override
+    protected AbstractAccessTokenRequest buildObtainRequest(WebClient accessTokenEndpoint,
+                                                            String accessTokenEndpointPath,
+                                                            @Nullable ClientAuthorization clientAuthorization) {
+        return new ClientCredentialsTokenRequest(accessTokenEndpoint, accessTokenEndpointPath,
+                                                 // clientAuthorization require for this Grant flow
+                                                 requireNonNull(clientAuthorization, "clientAuthorization"));
```

Comment:
Global comment: How about accepting the required properties at the construction time of the builder? e.g. `OAuth2ClientCredentialsGrant.builder(...., ClientAuthorization)`? This prevents a user from forgetting to set a required property ever.

Reply:
It's a bit tricky. For instance, clientAuthorization is required for OAuth2ClientCredentialsGrant but there are several different ways to build it (using authorizationSupplier or credentialsSupplier).